src/client/handlers/VaultsSecretsMkdir.ts

```diff
@@ -2,50 +2,60 @@ import type { DB } from '@matrixai/db';
 import type {
   ClientRPCRequestParams,
   ClientRPCResponseResult,
-  SecretMkdirMessage,
-  SuccessMessage,
+  SecretDirMessage,
+  SuccessWithErrorMessage,
 } from '../types';
 import type VaultManager from '../../vaults/VaultManager';
-import { UnaryHandler } from '@matrixai/rpc';
+import { DuplexHandler } from '@matrixai/rpc';
 import * as vaultsUtils from '../../vaults/utils';
 import * as vaultsErrors from '../../vaults/errors';
 import * as vaultOps from '../../vaults/VaultOps';
 
-class VaultsSecretsMkdir extends UnaryHandler<
+class VaultsSecretsMkdir extends DuplexHandler<
   {
     vaultManager: VaultManager;
     db: DB;
   },
-  ClientRPCRequestParams<SecretMkdirMessage>,
-  ClientRPCResponseResult<SuccessMessage>
+  ClientRPCRequestParams<SecretDirMessage>,
+  ClientRPCResponseResult<SuccessWithErrorMessage>
 > {
-  public handle = async (
-    input: ClientRPCRequestParams<SecretMkdirMessage>,
-  ): Promise<ClientRPCResponseResult<SuccessMessage>> => {
-    const { vaultManager, db } = this.container;
-    await db.withTransactionF(async (tran) => {
-      const vaultIdFromName = await vaultManager.getVaultId(
-        input.nameOrId,
-        tran,
-      );
-      const vaultId =
-        vaultIdFromName ?? vaultsUtils.decodeVaultId(input.nameOrId);
-      if (vaultId == null) {
-        throw new vaultsErrors.ErrorVaultsVaultUndefined();
-      }
-      await vaultManager.withVaults(
-        [vaultId],
-        async (vault) => {
-          await vaultOps.mkdir(vault, input.dirName, {
-            recursive: input.recursive,
-          });
-        },
-        tran,
-      );
-    });
-    return {
-      success: true,
-    };
+  public handle = async function* (
+    input: AsyncIterable<ClientRPCRequestParams<SecretDirMessage>>,
+  ): AsyncGenerator<ClientRPCResponseResult<SuccessWithErrorMessage>> {
+    const { vaultManager, db }: { vaultManager: VaultManager; db: DB } =
+      this.container;
+    let metadata: any = undefined;
+    yield* db.withTransactionG(
+      async function* (tran): AsyncGenerator<SuccessWithErrorMessage> {
+        for await (const secretDirMessage of input) {
+          // Unpack input
+          if (metadata == null) metadata = secretDirMessage.metadata ?? {};
+          const nameOrId = secretDirMessage.nameOrId;
+          const dirName = secretDirMessage.dirName;
+          // Get vaultId
+          const vaultIdFromName = await vaultManager.getVaultId(nameOrId, tran);
+          const vaultId =
+            vaultIdFromName ?? vaultsUtils.decodeVaultId(nameOrId);
+          if (vaultId == null) {
+            throw new vaultsErrors.ErrorVaultsVaultUndefined();
+          }
+          // Write directories. This doesn't need to be grouped by vault names,
+          // as no commit is created for empty directories anyways.
+          let response: SuccessWithErrorMessage = { success: false };
```

Comment:
My LSP was throwing an error if I assigned `undefined` when assigning a value, so I assigned a dummy value as it would be overridden in the `withVaults` bit anyways. Should this be changed, or is it fine as-is?

Reply:
You can do `let response: type | undefined;` if you want it to start unininialized. I think `let response: Type;` *should* also be valid.

Reply:
Technically, I can just do `let response: type;`, but my LSP still complains, saying that the variable is potentially unbound. The LSP I am using is the same one used by VS Code, so this issue isn't specific to my LSP. I have done the following to tell my LSP that I know better: `if (response!.success) {...}`.